### ISSUE TITLE
Wishlist: support for using spyder from one virtual env to develp an app in another virtual env

### ISSUE THREAD

**Issue by callegar** [spyder-ide/spyder#16359]
In the development of an application I find spyder's own dependencies somehow reducing flexibility in managing the dependencies of an application under development when using python virtual environments.

This is because, to the best of my understanding, to use the spyder IDE to assist in the development of an application, it is necessary to install spyder in the same virtual env as the application under development, and spyder brings in its own relatively large set of dependencies with their own interrelations.

If I understand correctly, it is — in principle — possible to use spyder from a virtual environment to develop in another virtual environment, but this requires configuring spyder to use the python interpreter from this other virtual environment. Unfortunately, I also understand that the interpreter configuration option is a global one, so that if spyder is used for multiple tasks one needs to either configure the option back and forward or to wrap the invocation of spyder in a script so that its configuration is seeked in the virtual environment of the software-in-development.

At least this is what I get from https://stackoverflow.com/questions/44389630/using-spyder-with-virtualenv where a wrapping script is proposed to assure that spyder gets a different spyder.ini file for each virtual environment.

I wonder if it might be possible and useful to enable the per-project configuration of an interpreter to avoid these burdens.

Please take my apology in case I have a wrong understanding of the matter.

### Issue Report Checklist

* [X] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [X] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [X] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

**Comment by callegar** [comment]
I now see that there is an extensive discussion about something that might be related to my enhancement suggestion in https://github.com/spyder-ide/ux-improvements/issues/10.

However, I would like to suggest the following points for consideration:
- the indication of the interpreter name (as it is now in the status bar) seems insufficient to me to disambiguate the environments. In my case I always see "custom (Python 3.8.10)" because most of my virtual environments are using that even if the virtual envs are completely different in the package assortment they offer.  The problem here is that with virtual env, when the interpreter is called, this brings you in the virtual environment (e.g. calling python3.8 from env1/bin brings you in env1, while calling python3.8 from env2/bin brings you in env2).
- this seems to me something sensible to link to projects.

Should I propose these points in https://github.com/spyder-ide/ux-improvements/issues/10?

**Comment by ccordoba12** [comment]
Hey @callegar, thanks for reporting. You said:

> This is because, to the best of my understanding, to use the spyder IDE to assist in the development of an application, it is necessary to install spyder in the same virtual env as the application under development

This is not correct. Although it's not easy at the moment, you can switch envs from a single Spyder installation by following the instructions mentioned here:

http://docs.spyder-ide.org/current/faq.html#using-existing-environment

> so that if spyder is used for multiple tasks one needs to either configure the option back and forward or to wrap the invocation of spyder in a script so that its configuration is seeked in the virtual environment of the software-in-development.

Correct. This should be solved once we add a way configure options per project (see issue #9804).

> At least this is what I get from https://stackoverflow.com/questions/44389630/using-spyder-with-virtualenv where a wrapping script is proposed to assure that spyder gets a different spyder.ini file for each virtual environment.

You can use the `--conf-dir` command line option (added in our 5.1.2 version) to have different configuration options per project now. That's the best you can do until we can properly solve issue #9804.

**Comment by ccordoba12** [comment]
You also said:

> the indication of the interpreter name (as it is now in the status bar) seems insufficient to me to disambiguate the environments

The thing is at the moment we only support conda and pyenv envs. The rest appears as `custom`. What kind of virtualenvs are you using?

> Should I propose these points in spyder-ide/ux-improvements#10?

No, thanks. It's not necessary because we're aware of these limitations and they need to be solved here.

**Comment by callegar** [comment]
@ccordoba12 thanks for the extensive explanations...

I am using just `virtualenv` to create the virtual environments... and I see that there is no concept of *environment name` in it.  But I wonder if in place of *custom* it may be useful to write something else...  Some possibilities could include:
- providing a label to go with the interpreter in the interpreter configuration dialog (this would be particularly useful if the interpreted configuration is made per-project);
- checking if the interpreter path has the form `<dir>/bin/python...` since in this case the `dir` component may possibly be used in place of `custom`.

**Comment by callegar** [comment]
Now trying the following:
- I have a "development" virtual env with the development IDE Spyder
- I have a "project" virtual env with the dependency of the project under development

Because of how virtualenv works, if I launch `development/bin/spyder` then the system automatically switches to the development virtual environment and launches spyder. So far so good.

Now, if I use `development/bin/spyder --conf-dir project/etc/spyder` I can have a configuration setup of spyder *private* for my development project as suggested. In it, I am setting the interpreter to `project/bin/python3`.  Because of how the virtual environment work, this means that the interpreter should automatically use packages from the project repo.

However this breaks spyder that at this point cannot find spyder-kernel anymore and cannot start any console.

Any clue?

**Comment by ccordoba12** [comment]
> checking if the interpreter path has the form <dir>/bin/python...

Yeah, that's a good idea. However, we're really swamped at the moment, so this will have to wait several months to be implemented, perhaps a year.

> However this breaks spyder because at this point it cannot find spyder-kernel anymore and cannot start any console.

As it's explained in the FAQ I mentioned above:

http://docs.spyder-ide.org/current/faq.html#using-existing-environment

you need to install spyder-kernels in any environment you want to use with Spyder with.

> Is it expected that if I set spyder to use an interpreter from a virtual environment it cannot find its own components anymore? Is it a bug? Any clue?

Yes, it is expected and no, it's not a bug. Spyder-kernels is necessary for the Variable explorer, debugger and Plots to work.
